common/src/primitives/merkle/tree/mod.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod tree_size;
+
+use std::num::NonZeroUsize;
+
+use crypto::hash::StreamHasher;
+
+use crate::primitives::{
+    id::{default_hash, DefaultHashAlgoStream},
+    H256,
+};
+
+use self::tree_size::TreeSize;
+
+use super::{pos::NodePosition, MerkleTreeAccessError, MerkleTreeFormError};
+
+pub enum AdjacentLeavesIndices {
+    Alone(usize),
+    Together(usize, usize),
+}
+
+/// Merkle tree in the form of a vector, where the bottom leaves are the based, and the root is
+/// the last element.
+/// Definitions:
+/// - Leaf: A node that cannot have children.
```

Comment:
cannot or does not? I'd argue it's the definition rather than a constraint

Reply:
That's incredibly picky though so I'm more than happy for you to ignore me entirely

Reply:
Fixed. I don't really care which we use, it's just language gymnastics on assertion.